requirements.txt

```diff
@@ -3,6 +3,7 @@
 # Downgrade the setuptools version to workaround the issue replacing the setuptools dependency
 # IOError METADATA no such file. Check https://github.com/pypa/setuptools/issues/951 for more details.
 setuptools==33.1.1
+apscheduler==3.5.1
```

Comment:
yeah. Me and @cmcginty briefly discussed about `apscheduler` being a overkill, but it seemed pretty straight forward to set up and use. I was hoping to use their native `TornadoScheduler`, but the current tornado version we use doesn't support it. I'll check out the `sched` module. Also good point on using safe thread. I'll verify the behavior on uncaught exceptions.

Reply:
As for apscheduler being easy to set up and use -- I totally agree. The code itself that you had around it is pretty clean. I'm less concerned about that than just the size and functionality of the library that we won't (and might never) leverage.

I don't think using `sched` would be much more complex. Something like:
```python
def __init__(self):
    self._hb_scheduler = sched.scheduler()
    SafeThread(target=self._start_beatin, name='HeartbeatLoop', daemon=True).start()  # this could also go in a public start() method similar to BuildRequestHandler

def _start_beatin(self):
    self._hb_scheduler.enter(0, 0, self._do_heartbeat)
    self._hb_scheduler.run()

def heartbeat_loop():
    # network request calls, error handling, metrics recording here
    self._hb_scheduler.enter(self._hb_freq, 0, self._do_heartbeat)  # schedule next beat. (if you wanted to be super precise you could subtract from the frequency the amount of time this method took to execute, but that shouldn't matter much for the intervals we're talking about.)
```

Reply:
uhh.. I hate how GitHub handles comments, but I reimplemented the heartbeat using sched. I still not fully satisfied, but I think it's better than before